cpp/modmesh/buffer/pymod/wrap_SimpleArrayPlex.cpp

```diff
@@ -28,28 +28,32 @@
 
 #include <modmesh/buffer/pymod/buffer_pymod.hpp> // Must be the first include.
 
+#include <type_traits>
+
 namespace modmesh
 {
 
 namespace python
 {
 
 /// Execute the callback function with the typed array
+/// @tparam A either `SimpleArrayPlex` or `const SimpleArrayPlex`
 /// @tparam C the type of the callback function
 /// @param arrayplex the plex array, which is the wrapper of the typed array
-/// @param callback the callback function, which has the typed array as the argument
+/// @param callback the callback function, which has the mutable typed array as the argument
 /// @return the return type of the callback function
-template <typename C>
+template <typename A, typename C, typename = std::enable_if_t<std::is_same_v<std::remove_const_t<A>, SimpleArrayPlex>>>
 // NOLINTNEXTLINE(misc-use-anonymous-namespace)
-static auto execute_callback_with_typed_array(SimpleArrayPlex & arrayplex, C && callback)
+static auto execute_callback_with_typed_array(A & arrayplex, C && callback)
 {
 // We get the typed array from the arrayplex and call the callback function with the typed array.
-#define DECL_MM_RUN_CALLBACK_WITH_TYPED_ARRAY(DataType, ArrayType)                 \
-    case DataType:                                                                 \
-    {                                                                              \
-        auto * array = static_cast<ArrayType *>(arrayplex.mutable_instance_ptr()); \
-        return callback(*array);                                                   \
-        break;                                                                     \
+#define DECL_MM_RUN_CALLBACK_WITH_TYPED_ARRAY(DataType, ArrayType)                                                     \
+    case DataType:                                                                                                     \
+    {                                                                                                                  \
+        using ArrayTypePtr = typename std::conditional<std::is_const<A>::value, const ArrayType *, ArrayType *>::type; \
```

Comment:
To match your use of `remove_const_t`, you could try to change it to `add_pointer` (https://en.cppreference.com/w/cpp/types/add_pointer).